File: src/tinychain/transaction_pool.py
class TransactionPool:
    def __init__(self, max_size):
        self.transactions = {}
        self.max_size = max_size

    def add_transaction(self, transaction):
        if len(self.transactions) >= self.max_size:
            raise ValueError("Transaction pool is full")
        if len(self.transactions) < self.max_size:
            self.transactions[transaction.transaction_hash] = transaction

    def remove_transaction(self, transaction):
        self.transactions.pop(transaction.transaction_hash, None)

    def get_transactions(self):
        return sorted(self.transactions.values(), key=lambda tx: tx.fee, reverse=True)

    def get_transaction_by_hash(self, hash):
        return self.transactions[hash]

    def is_empty(self):
        return not self.transactions

File: src/tinychain-rust/src/transaction_pool.rs
pub struct TransactionPool {
    transactions: Vec<Transaction>,
    max_size: usize,
}

impl TransactionPool {
    pub fn new(max_size: usize) -> Self {
        TransactionPool {
            transactions: Vec::new(),
            max_size,
        }
    }

    pub fn add_transaction(&mut self, transaction: Transaction) -> Result<(), &'static str> {
        if self.transactions.len() >= self.max_size {
            return Err("Transaction pool is full");
        }
        self.transactions.push(transaction);
        Ok(())
    }

    pub fn remove_transaction(&mut self, transaction_hash: &str) {
        self.transactions.retain(|tx| tx.transaction_hash != transaction_hash);
    }

    pub fn get_transactions(&self) -> Vec<&Transaction> {
        let mut sorted_transactions: Vec<&Transaction> = self.transactions.iter().collect();
        sorted_transactions.sort_by(|a, b| b.fee.cmp(&a.fee));
        sorted_transactions
    }

    pub fn get_transaction_by_hash(&self, hash: &str) -> Option<&Transaction> {
        self.transactions.iter().find(|tx| tx.transaction_hash == hash)
    }

    pub fn is_empty(&self) -> bool {
        self.transactions.is_empty()
    }
}
